File: python/bc-lifehash/src/bc_lifehash/_hsb_color.py
from __future__ import annotations

import math

from ._color import Color, _to_f32, clamped, modulo


class HSBColor:
    """HSB/HSV color with conversion to RGB."""

    __slots__ = ("hue", "saturation", "brightness")

    def __init__(self, hue: float, saturation: float, brightness: float) -> None:
        self.hue = hue
        self.saturation = saturation
        self.brightness = brightness

    @classmethod
    def from_hue(cls, hue: float) -> HSBColor:
        return cls(hue, 1.0, 1.0)

    def color(self) -> Color:
        v = clamped(self.brightness)
        s = clamped(self.saturation)

        if s <= 0.0:
            return Color(v, v, v)

        h = modulo(self.hue, 1.0)
        if h < 0.0:
            h += 1.0
        h *= 6.0
        # C++ uses floorf (f32 precision)
        i = int(math.floor(_to_f32(h)))
        f = h - float(i)
        p = v * (1.0 - s)
        q = v * (1.0 - s * f)
        t = v * (1.0 - s * (1.0 - f))

        if i == 0:
            return Color(v, t, p)
        elif i == 1:
            return Color(q, v, p)
        elif i == 2:
            return Color(p, v, t)
        elif i == 3:
            return Color(p, q, v)
        elif i == 4:
            return Color(t, p, v)
        elif i == 5:
            return Color(v, p, q)
        else:
            raise RuntimeError("Internal error in HSB conversion")

File: typescript/bc-lifehash/src/hsb-color.ts
import { Color, clamped, modulo } from './color.js';

export class HSBColor {
    readonly hue: number;
    readonly saturation: number;
    readonly brightness: number;

    constructor(hue: number, saturation: number, brightness: number) {
        this.hue = hue;
        this.saturation = saturation;
        this.brightness = brightness;
    }

    static fromHue(hue: number): HSBColor {
        return new HSBColor(hue, 1.0, 1.0);
    }

    color(): Color {
        const v = clamped(this.brightness);
        const s = clamped(this.saturation);

        if (s <= 0.0) {
            return new Color(v, v, v);
        }

        let h = modulo(this.hue, 1.0);
        if (h < 0.0) {
            h += 1.0;
        }
        h *= 6.0;
        // C++ uses floorf (f32 precision)
        const i = Math.floor(Math.fround(h));
        const f = h - i;
        const p = v * (1.0 - s);
        const q = v * (1.0 - s * f);
        const t = v * (1.0 - s * (1.0 - f));

        switch (i) {
            case 0: return new Color(v, t, p);
            case 1: return new Color(q, v, p);
            case 2: return new Color(p, v, t);
            case 3: return new Color(p, q, v);
            case 4: return new Color(t, p, v);
            case 5: return new Color(v, p, q);
            default: throw new Error('Internal error in HSB conversion');
        }
    }
}
